backends/gstreamer/source-element/src/source.rs

```diff
@@ -0,0 +1,174 @@
+    inner_appsrc_proxy!(set_callbacks, callbacks, AppSrcCallbacks, ());
+    inner_appsrc_proxy!(set_property_format, format, gst::Format, ());
+    inner_appsrc_proxy!(set_size, size, i64, ());
+    inner_appsrc_proxy!(set_stream_type, type_, AppStreamType, ());
+}
+
+impl ObjectImpl<BaseSrc> for ServoSrc { }
+
+impl ElementImpl<BaseSrc> for ServoSrc { }
+
+impl URIHandlerImpl for ServoSrc {
+    fn get_uri(&self, _element: &gst::URIHandler) -> Option<String> {
+        Some("servosrc://".to_string())
+    }
+
+    fn set_uri(&self, _element: &gst::URIHandler, _uri: Option<String>) -> Result<(), glib::Error> {
+        Ok(())
+    }
+}
+
+impl BaseSrcImpl<BaseSrc> for ServoSrc {
+    fn query(&self, src: &BaseSrc, query: &mut gst::QueryRef) -> bool {
+        // In order to make buffering/downloading work as we want, apart from
+        // setting the appropriate flags, the source needs to either:
+        //
+        // 1. be an http, mms, etc. scheme
+        // 2. report that it is "bandwidth limited".
+        //
+        // 1. is not straightforward because we are using a servosrc scheme for now.
+        // This may change in the future if we end up implementing something
+        // like a custom `servohttpsrc` or wrapping the existing servosrc
+        // in a bin src that implements http/https/data URI handlers, which
+        // is what WebKit does.
+        //
+        // For 2. we need to make servosrc handle the scheduling properties query
+        // to report that it "is bandwidth limited".
+        if let gst::QueryView::Scheduling(ref mut query) = query.view_mut() {
+            query.set(
+                gst::SchedulingFlags::SEQUENTIAL |
+                gst::SchedulingFlags::BANDWIDTH_LIMITED,
+                1, -1, 0
+            );
+            query.add_scheduling_modes(&[gst::PadMode::Push]);
+        }
+
+        BaseSrcBase::parent_query(src, query)
+    }
+}
+
+struct ServoSrcStatic;
+
+impl ImplTypeStatic<BaseSrc> for ServoSrcStatic {
+    fn get_name(&self) -> &str {
+        "ServoSrc"
+    }
+
+    fn new(&self, element: &BaseSrc) -> Box<BaseSrcImpl<BaseSrc>> {
+        ServoSrc::new(element)
+    }
+
+    fn class_init(&self, klass: &mut BaseSrcClass) {
+        ServoSrc::class_init(klass);
+    }
+
+    fn type_init(&self, token: &TypeInitToken, type_: glib::Type) {
+        register_uri_handler(token, type_, self);
+    }
+}
+
+impl URIHandlerImplStatic<BaseSrc> for ServoSrcStatic {
+    fn get_impl<'a>(&self, imp: &'a Box<BaseSrcImpl<BaseSrc>>) -> &'a URIHandlerImpl {
+        imp.downcast_ref::<ServoSrc>().unwrap()
+    }
+
+    fn get_type(&self) -> gst::URIType {
+        gst::URIType::Src
+    }
+
+    fn get_protocols(&self) -> Vec<String> {
+        vec!["servosrc".into()]
+    }
+}
+
+pub fn register(plugin: &gst::Plugin) -> bool {
+    let type_ = register_type(ServoSrcStatic);
+    gst::Element::register(plugin, "servosrc", 0, type_)
```

Comment:
I don't know for sure, but I guess we can argue if it's required to register an element that will be used only internally in a project and not exposed to the general gstreamer setup

Reply:
You need to register it so that the URI handler implementation of it is found